### DISCUSSION TITLE
Parent <-> child communication design.

### DISCUSSION THREAD

**Discussion by slimshader** [microsoft/vs-streamjsonrpc#1025] [Q&A] [answered]
Hi,

My case: launcher application that is able to spawn 2 child apps. I want my launcher and child apps be able to talk to each other via json rpc, meaning sending multiple messages back and forth when needed. Based on od cond sampes I implemented it this way:

1. Launcher creates a server based on InOut NamedPipeServerStream with name "Parent"
2. Launcher starts child app
3. child app creates InOut server with generated name
4. child creates NamedPipeClientStream to "Parent"
5. child calls remote procedure on parent "ChildReady" with generated name as paramter
6. Parent creates NPCS with child's name
7. parent and child send messages to each other via respecitve Client streams
8. server streams get recreated after each precessed message (via while (true) loops)
```csharp
            while (true)
            {
                var stream = new NamedPipeServerStream("StreamJsonRpcSamplePipe", PipeDirection.InOut, NamedPipeServerStream.MaxAllowedServerInstances, PipeTransmissionMode.Byte, PipeOptions.Asynchronous);

                await stream.WaitForConnectionAsync();

                var jsonRpc = JsonRpc.Attach(stream, this);

                await jsonRpc.Completion;
            }
```

is that the way to go for this case?

**Comment by AArnott** [comment, marked as answer]
Named pipes are bidirectional, so you don't need each process to create both a server and a client pipe. The parent process should create a named pipe and all child processes can connect to it as clients. Now each client has a bidirectional stream with the parent, over which JSON-RPC can be done.
The server would indeed have a loop that would repeatedly listen on the named pipe for incoming connections and as each connection is made, it should connect a JsonRpc object to it using Attach as you've done. But it should _not_ await jsonRpc.Completion or else it'll hang there until the client hangs up before listening for the next incoming connection.

Once the connection is established, you can and should use it for all your JSON-RPC messages. Don't recreate the stream after each processed message.

Here is a sample named pipe server that can accept multiple clients:
https://github.com/microsoft/vs-servicehub/blob/main/src/Microsoft.ServiceHub.Framework/ServerFactory.cs
